### PR TITLE
Refactor optimize() functions to take a CompileOptions

### PR DESCRIPTION
**Pull request by nickgg**
*Description*: As a follow up #2451, push the CompileOptions struct down into the Optimize functions as well for future expansion of optimization configuration. I kept the existing CompilationMode function to ease transition.
*Testing*: unit tests, w/ ASAN.
*Documentation*:
This changes the Backend interface and will require updating any new backends.

### CODE REVIEW

File: include/glow/Backends/CompileOptions.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+namespace glow {
+
+enum class CompilationMode {
+  Train, /// Compile the graph in preperation for training.
+  Infer, /// Compile the graph for inference. Notice that this operation
+         /// changes the graph in a way that is not reversible.
+};
+
+/// Configuration options for the compile() method on Backend.
+struct CompileOptions {
```
Comment (opti-mix) [inline include/glow/Backends/CompileOptions.h]: We have `CompilationMode`, but `CompileOptions`. This looks odd. Should we use a more consistent naming?

Comment (opti-mix) [reply include/glow/Backends/CompileOptions.h]: Nit: Why do we have this header in the Backends subdirectory? Is it the most logical place to put it?

**Comment by jfix71** [comment]
Maybe we should consider the same builder pattern as we use in NodeGen/InstrGen -- I think it would make things cleaner. Specifically adding something to the CompileOptions struct like:
```cpp
CompileOptions &setCompilationMode(CompilationMode m) {
  mode = m;
  return *this;
}
```

Then you can do something like:
```cpp
EE_.compile(F, CompileOptions()
                   .setCompilationMode(CompilationMode::Infer)
                   .setCollectConstants(true));
```

File: include/glow/Backends/CompileOptions.h (hunk 1 above)
Comment (nickgg) [reply include/glow/Backends/CompileOptions.h]: What's your preferred alternative?

Comment (opti-mix) [reply include/glow/Backends/CompileOptions.h]: For naming, I'd go with `CompilationMode` and `CompilationOptions`.

For the subdirectory, I don't have any strong opinion. May be a new `include/glow/Options/` subdirectory?

**Comment by nickgg** [comment]
I'm not totally convinced by the builder pattern here but also not really against it. Can we gather a few more opinions on it?

File: include/glow/Backends/CompileOptions.h (hunk 1 above)
Comment (nickgg) [reply include/glow/Backends/CompileOptions.h]: CompilationOptions sounds fine to me but creating a new subdirectory for a single header seems overkill, what else would we need to put in there?

**Comment by opti-mix** [comment]
> I'm not totally convinced by the builder pattern here but also not really against it.

+1.
I'm also not totally convinced if using the builder pattern for options is worth the effort.

File: include/glow/Backends/CompileOptions.h (hunk 1 above)
Comment (opti-mix) [reply include/glow/Backends/CompileOptions.h]: > but creating a new subdirectory for a single header seems overkill, what else would we need to put in there?

Yeah, I agree. My initial (very small) concern was that the options are not for backends only. They are also used by optimizers, etc.

**Comment by nickgg** [comment]
@jfix71 I'll land this now to avoid conflicts, and we can come back to the builder pattern question.

**Pull request merged**